### DISCUSSION TITLE
Question about exercise 4.2

### DISCUSSION THREAD

**Discussion by bonosa** [Lightning-AI/dl-fundamentals#21] [Exercises] [answered]
Hi Sebastian,

the only thing I can think of in getitem after you normalize the image is to resize it. Not near my desk so I haven’t tried it. Am I thinking correctly? How will I know if that’s the missing code and why is the resize argument 32?  Sorry if it’s a dumb question 🤔

**Comment by rasbt** [comment]
Thanks for asking, I am happy to chat about this, so no need to apologize :).

Regarding the resize argument, that's not necessary. You mean in https://github.com/Lightning-AI/dl-fundamentals/blob/main/unit04-multilayer-nets/exercises/2_fashion-mnist/exercise_2_fashion-mnist.ipynb right? So here you don't need to resize. The images are 28x28 images. 

You probably meant the following line?
    
    img = torch.tensor(img).to(torch.float32)

Here, we are casting the images to a float array. By default, image data is an integer between 0 and 255. We usually need float variables for computing the gradients etc.

**Reply by bonosa** [reply]
What is the missing line of code after img/255? Thanks for replying. Re exercise 2. In unit 4

**Reply by rasbt** [reply]
As a little tip, you would need to return both the image (`img`) and the corresponding class label. So what you need is 

1. a line to get the class label (`label`)
2. a line at the bottom that returns both `img` and `label`

**Reply by Baburam208** [reply]
Dear Sebastian,


```python
class MyDataset(Dataset):
    def __init__(self, img, label, transform=None):
        self.img = torch.tensor(img, dtype=torch.float32)
        self.label = torch.tensor(label, dtype=torch.float32)
        self.transform = transform
        
    def __getitem__(self, index):
        img = self.img[index]
        img = img/255.
        y = self.label[index]
        
        if transform is not None:
            img = self.transform(img)
            
        return img, label
    
    def __len__(self):
        return self.labels.shape[0]
    
# (say x_train, y_train) be training dataset
train_ds = MyDataset(x_train, y_train)
```

Is this the correct one or way you expect.

**Reply by rasbt** [reply]
Hi @Baburam208 It would depend a bit on what `img` and `label` are that you are passing to the class constructor. To make your code work, these would have to be arrays of all the images and labels. Nothing wrong with that btw., just saying you have to make sure that these are the right inputs.

**Comment by bonosa** [comment]
Aaargh I don’t know enough python to know what %kind means in the os.path.join. Also is there only one argument passed into get item? And only one line of code in the… placeholders?

**Reply by rasbt** [reply]
Oh yeah, maybe should have used f-strings. This (`%s`) is an old Python way of substituting in a string. E.g.

```python
In [7]: path = "my-path"
In [8]: kind = "train"
In [9]: os.path.join(path, '%s-images-idx3-ubyte.gz' % kind)
Out[9]: 'my-path/train-images-idx3-ubyte.gz'
```

**Reply by bonosa** [reply]
Rings a bell now that you say it! 😂thanks

**Reply by intelligencethink** [reply]
Hello Sebastian:

Thank you for the fundamental deep learning video.

I also want to ask something about exercise_2_fashion-mnist.
I completed the MyDataset missing blank space, and I follow unit 4.4 _video2 to check the first three batches. When I continue to do step 4) The training loop,  at the line loss = F.cross_entropy (logits, labels), I get the error message.
Runtime Error; expected scalar type long but found Byte. Do I need to any data transform to convert datatype?

I use load_fasionmnist inside the __init__ to return the images and labels, and add the lines to get img and label in the__getitm__ to return img, label

**Reply by rasbt** [reply]
hmm, when you print the batches, maybe check the data types of the features and the labels. E.g.,

```python
print("features", img.dtype)
print("label", label.dtype)
```

`features` should be type `float32` and `label` should be type `int64`

If the types don't match these, you can recast them via `label.to(torch.int64)` etc.

**Reply by intelligencethink** [reply]
Thank you very much for the reply.

I check the data type of the label. It is int8.
Following the syntax, I convert the label datatype and solve the error.
I got a training accuracy of 87.18% on 5,5000 training data just using the parameters provided.

Thanks very much for your kind tutorials.

**Reply by rasbt** [reply]
Nice, glad to hear that you got it to work, and the accuracy looks good!

**Comment by bonosa** [comment, marked as answer]
Thanks for the way forward. I was kinda lost 👋
